pkg/util/fxutil/args.go

```diff
@@ -0,0 +1,80 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package fxutil
+
+import (
+	"reflect"
+
+	"go.uber.org/fx"
+)
+
+var errorInterface = reflect.TypeOf((*error)(nil)).Elem()
+
+// delayedFxInvocation delays execution of a function, while allowing
+// Fx to provide its arguments.
+type delayedFxInvocation struct {
```

Comment:
I have to admit I have read through the docs part somewhat diagonally, so I may have missed what might be a complete explanation. Can we outline some example of cases where delayed execution will be necessary?

Reply:
Exactly.  And Start ordering is a partial order, so the only way to say "run this Start hook after all other Start hooks" is to depend on _all_ other components, which is impractical.

This is an implementation detail of fxutil (and in particular `fxutil.OneShot` and `fxutil.Test`, both of which want to call a function with a "fully-started" app),.  It may be that later versions of fx provide better support for this, such as with an `fx.AfterStart(..)` or `Hooks#AfterStart`, in which case we can switch to that support.